Mathlib/Computability/AkraBazzi/AkraBazzi.lean

```diff
@@ -0,0 +1,1485 @@
+  refine ⟨c, ⟨hc_pos, hc_lt_one⟩, ?_⟩
+  have hlo := isLittleO_self_div_log_id
+  rw [Asymptotics.isLittleO_iff] at hlo
+  have hlo' := hlo h₁
+  filter_upwards [hlo', R.eventually_r_le_b] with n hn hn'
+  intro i
+  rw [Real.norm_of_nonneg (by positivity)] at hn
+  simp only [Real.norm_of_nonneg (by positivity : 0 ≤ (n : ℝ))] at hn
+  calc r i n ≤ b i * n + n / log n ^ 2 := by exact hn' i
+             _ ≤ b i * n + (1 - b R.max_bi) / 2 * n := by gcongr
+             _ = (b i + (1 - b R.max_bi) / 2) * n := by ring
+             _ ≤ (b R.max_bi + (1 - b R.max_bi) / 2) * n := by gcongr; exact R.max_bi_le _
+
+lemma eventually_r_pos : ∀ᶠ (n:ℕ) in atTop, ∀ i, 0 < r i n := by
+  rw [Filter.eventually_all]
+  exact fun i => (R.tendsto_atTop_r i).eventually_gt_atTop 0
+
+lemma eventually_log_b_mul_pos : ∀ᶠ (n:ℕ) in atTop, ∀ i, 0 < log (b i * n) := by
+  rw [Filter.eventually_all]
+  intro i
+  have h : Tendsto (fun (n:ℕ) => log (b i * n)) atTop atTop :=
+    Tendsto.comp tendsto_log_atTop
+      <| Tendsto.const_mul_atTop (b_pos R i) tendsto_nat_cast_atTop_atTop
+  exact h.eventually_gt_atTop 0
+
+lemma fin_nonempty {R : AkraBazziRecurrence T g a b r} : Nonempty (Fin k) := by
+  rw [←Fin.pos_iff_nonempty]
+  exact R.k_gt_zero
+
+lemma fin_univ_nonempty {R : AkraBazziRecurrence T g a b r} :
+    Finset.Nonempty (univ : Finset <| Fin k) := by
+  rw [Finset.univ_nonempty_iff]
+  exact R.fin_nonempty
+
+@[aesop safe apply] lemma T_pos (n : ℕ) : 0 < T n := by
+  induction n using Nat.strongInductionOn
+  case ind n h_ind =>
+    rcases lt_or_le n R.n₀ with hn|hn
+    case inl => exact R.T_gt_zero' n hn   -- n < R.n₀
+    case inr =>   -- R.n₀ ≤ n
+      rw [R.h_rec n hn]
+      have := R.g_nonneg
+      refine add_pos_of_pos_of_nonneg (Finset.sum_pos ?sum_elems R.fin_univ_nonempty) (by aesop)
+      exact fun i _ => mul_pos (R.a_pos i) <| h_ind _ (R.r_lt_n i _ hn)
+
+@[aesop safe apply]
+lemma T_nonneg (n : ℕ) : 0 ≤ T n := le_of_lt <| R.T_pos n
+
+/-!
+#### Smoothing function
+
+We define `ε` as the "smoothing function" `fun n => 1 / log n`, which will be used in the form of a
+factor of `1 ± ε n` needed to make the induction step go through.
+
+This is its own definition to make it easier to switch to a different smoothing function.
+For example, choosing `1 / log n ^ δ` for a suitable choice of `δ` leads to a slightly tighter
+theorem at the price of a more complicated proof.
+
+This part of the file then proves several properties of this function that will be needed later in
+the proof.
+-/
+
+/-- The "smoothing function" is defined as `1 / log n`. This is defined as an `ℝ → ℝ` function
+as opposed to `ℕ → ℝ` since this is more convenient for the proof, where we need to e.g. take
+derivatives. -/
+noncomputable def smoothingFn (n : ℝ) : ℝ := 1 / log n
+
+local notation "ε" => smoothingFn
+
+lemma one_add_smoothingFn_le_two {x : ℝ} (hx : exp 1 ≤ x) : 1 + ε x ≤ 2 := by
+  simp only [smoothingFn, ←one_add_one_eq_two]
```

Comment:
```suggestion
  simp only [smoothingFn, ← one_add_one_eq_two]
```
There is a consensus that this is the preferred style. Could you adapt the PR here and elsewhere?